File: src-tauri/src/deeplink.rs
use tauri::{AppHandle, Manager, Runtime};
use serde::{Deserialize, Serialize};

/// Parsed deep link data
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DeepLinkPayload {
    /// The action type (chat, room, channel, server, invite, settings)
    pub action: String,
    /// The target ID (room id, channel id, server id, user id, etc.)
    pub target: Option<String>,
    /// Additional parameters
    pub params: std::collections::HashMap<String, String>,
}

/// Parse a hearth:// URL into a DeepLinkPayload
/// 
/// Supported formats:
/// - hearth://chat/:userId - Open DM with user
/// - hearth://room/:roomId - Open a room  
/// - hearth://channel/:channelId - Navigate to a specific channel
/// - hearth://server/:serverId - Navigate to a specific server
/// - hearth://server/:serverId/:channelId - Navigate to server + channel
/// - hearth://invite/:code - Accept an invite
/// - hearth://invite/:code?server=:serverId - Accept invite with server context
/// - hearth://settings - Open settings
/// - hearth://settings/:section - Open specific settings section
/// - hearth://call/:callId - Join a voice call
pub fn parse_deep_link(url: &str) -> Option<DeepLinkPayload> {
    let url = url.trim();
    
    // Must start with hearth://
    if !url.starts_with("hearth://") {
        return None;
    }

    let path = &url[9..]; // Remove "hearth://"
    let mut parts: Vec<&str> = path.split('?').collect();
    let path_part = parts.remove(0);
    
    // Parse query params
    let mut params = std::collections::HashMap::new();
    if !parts.is_empty() {
        let query = parts.join("?");
        for pair in query.split('&') {
            if let Some((key, value)) = pair.split_once('=') {
                params.insert(
                    urlencoding::decode(key).unwrap_or_default().to_string(),
                    urlencoding::decode(value).unwrap_or_default().to_string(),
                );
            }
        }
    }

    // Parse path
    let segments: Vec<&str> = path_part.split('/').filter(|s| !s.is_empty()).collect();
    
    if segments.is_empty() {
        return None;
    }

    let action = segments[0].to_string();
    let target = segments.get(1).map(|s| s.to_string());

    // Handle special case: server/:serverId/:channelId
    // The third segment becomes a "channel" param
    if action == "server" && segments.len() >= 3 {
        params.insert("channel".to_string(), segments[2].to_string());
    }

    Some(DeepLinkPayload {
        action,
        target,
        params,
    })
}

/// Handle a deep link by emitting to the frontend
pub fn handle_deep_link<R: Runtime>(app: &AppHandle<R>, url: &str) {
    if let Some(payload) = parse_deep_link(url) {
        log::info!("Handling deep link: {:?}", payload);
        
        // Show and focus the window
        if let Some(window) = app.get_webview_window("main") {
            let _ = window.show();
            let _ = window.set_focus();
            
            // Emit the deep link event to frontend
            let _ = window.emit("deeplink", payload);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_chat_link() {
        let payload = parse_deep_link("hearth://chat/user123").unwrap();
        assert_eq!(payload.action, "chat");
        assert_eq!(payload.target, Some("user123".to_string()));
    }

    #[test]
    fn test_parse_room_link() {
        let payload = parse_deep_link("hearth://room/abc-def-ghi").unwrap();
        assert_eq!(payload.action, "room");
        assert_eq!(payload.target, Some("abc-def-ghi".to_string()));
    }

    #[test]
    fn test_parse_channel_link() {
        let payload = parse_deep_link("hearth://channel/chan-123").unwrap();
        assert_eq!(payload.action, "channel");
        assert_eq!(payload.target, Some("chan-123".to_string()));
    }

    #[test]
    fn test_parse_server_link() {
        let payload = parse_deep_link("hearth://server/server-456").unwrap();
        assert_eq!(payload.action, "server");
        assert_eq!(payload.target, Some("server-456".to_string()));
    }

    #[test]
    fn test_parse_server_channel_link() {
        let payload = parse_deep_link("hearth://server/server-456/chan-789").unwrap();
        assert_eq!(payload.action, "server");
        assert_eq!(payload.target, Some("server-456".to_string()));
        assert_eq!(payload.params.get("channel"), Some(&"chan-789".to_string()));
    }

    #[test]
    fn test_parse_invite_link() {
        let payload = parse_deep_link("hearth://invite/ABCD1234?ref=email").unwrap();
        assert_eq!(payload.action, "invite");
        assert_eq!(payload.target, Some("ABCD1234".to_string()));
        assert_eq!(payload.params.get("ref"), Some(&"email".to_string()));
    }

    #[test]
    fn test_parse_invite_with_server() {
        let payload = parse_deep_link("hearth://invite/XYZ789?server=server-123").unwrap();
        assert_eq!(payload.action, "invite");
        assert_eq!(payload.target, Some("XYZ789".to_string()));
        assert_eq!(payload.params.get("server"), Some(&"server-123".to_string()));
    }

    #[test]
    fn test_parse_settings_link() {
        let payload = parse_deep_link("hearth://settings").unwrap();
        assert_eq!(payload.action, "settings");
        assert_eq!(payload.target, None);
    }

    #[test]
    fn test_parse_settings_section() {
        let payload = parse_deep_link("hearth://settings/notifications").unwrap();
        assert_eq!(payload.action, "settings");
        assert_eq!(payload.target, Some("notifications".to_string()));
    }

    #[test]
    fn test_parse_call_link() {
        let payload = parse_deep_link("hearth://call/call-abc-123").unwrap();
        assert_eq!(payload.action, "call");
        assert_eq!(payload.target, Some("call-abc-123".to_string()));
    }
}

File: src/lib/utils/deeplink.ts
/**
 * Deep link utilities for generating hearth:// URLs
 * 
 * These URLs can be shared externally and will open the Hearth desktop app
 * when clicked, navigating directly to the specified resource.
 */

const PROTOCOL = 'hearth';

/**
 * Generate a deep link URL for a DM conversation with a user
 * @param userId - The user ID to open a chat with
 */
export function createChatLink(userId: string): string {
  return `${PROTOCOL}://chat/${encodeURIComponent(userId)}`;
}

/**
 * Generate a deep link URL for a channel
 * @param channelId - The channel ID to navigate to
 */
export function createChannelLink(channelId: string): string {
  return `${PROTOCOL}://channel/${encodeURIComponent(channelId)}`;
}

/**
 * Generate a deep link URL for a server
 * @param serverId - The server ID to navigate to
 * @param channelId - Optional channel ID within the server
 */
export function createServerLink(serverId: string, channelId?: string): string {
  if (channelId) {
    return `${PROTOCOL}://server/${encodeURIComponent(serverId)}/${encodeURIComponent(channelId)}`;
  }
  return `${PROTOCOL}://server/${encodeURIComponent(serverId)}`;
}

/**
 * Generate a deep link URL for an invite
 * @param inviteCode - The invite code
 * @param serverId - Optional server ID for context
 * @param referrer - Optional referrer parameter for tracking
 */
export function createInviteLink(
  inviteCode: string, 
  options?: { serverId?: string; referrer?: string }
): string {
  const params = new URLSearchParams();
  
  if (options?.serverId) {
    params.set('server', options.serverId);
  }
  if (options?.referrer) {
    params.set('ref', options.referrer);
  }
  
  const queryString = params.toString();
  const base = `${PROTOCOL}://invite/${encodeURIComponent(inviteCode)}`;
  
  return queryString ? `${base}?${queryString}` : base;
}

/**
 * Generate a deep link URL for settings
 * @param section - Optional settings section (notifications, appearance, etc.)
 */
export function createSettingsLink(section?: string): string {
  if (section) {
    return `${PROTOCOL}://settings/${encodeURIComponent(section)}`;
  }
  return `${PROTOCOL}://settings`;
}

/**
 * Generate a deep link URL for joining a voice call
 * @param callId - The call/voice channel ID to join
 */
export function createCallLink(callId: string): string {
  return `${PROTOCOL}://call/${encodeURIComponent(callId)}`;
}

/**
 * Parse a deep link URL and extract its components
 * @param url - The hearth:// URL to parse
 * @returns Parsed components or null if invalid
 */
export function parseDeepLink(url: string): {
  action: string;
  target: string | null;
  params: Record<string, string>;
} | null {
  const trimmed = url.trim();
  
  if (!trimmed.startsWith(`${PROTOCOL}://`)) {
    return null;
  }
  
  const withoutProtocol = trimmed.slice(PROTOCOL.length + 3); // Remove "hearth://"
  const [pathPart, queryPart] = withoutProtocol.split('?');
  
  // Parse query params
  const params: Record<string, string> = {};
  if (queryPart) {
    const searchParams = new URLSearchParams(queryPart);
    searchParams.forEach((value, key) => {
      params[key] = value;
    });
  }
  
  // Parse path segments
  const segments = pathPart.split('/').filter(s => s.length > 0);
  
  if (segments.length === 0) {
    return null;
  }
  
  const action = decodeURIComponent(segments[0]);
  const target = segments[1] ? decodeURIComponent(segments[1]) : null;
  
  // Handle server/:serverId/:channelId case
  if (action === 'server' && segments.length >= 3) {
    params['channel'] = decodeURIComponent(segments[2]);
  }
  
  return { action, target, params };
}

/**
 * Check if a URL is a valid Hearth deep link
 * @param url - The URL to check
 */
export function isDeepLink(url: string): boolean {
  return url.trim().startsWith(`${PROTOCOL}://`);
}
